src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -0,0 +1,373 @@
+                }
+                return _length;
+            }
+        }
+
+        /// <summary>
+        /// Peeks at the next value without advancing the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryPeek(out T value)
+        {
+            if (_moreData)
+            {
+                value = CurrentSpan[CurrentSpanIndex];
+                return true;
+            }
+            else
+            {
+                value = default;
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Read the next value and advance the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryRead(out T value)
+        {
+            if (End)
+            {
+                value = default;
+                return false;
+            }
+
+            value = CurrentSpan[CurrentSpanIndex];
+            CurrentSpanIndex++;
+            Consumed++;
+
+            if (CurrentSpanIndex >= CurrentSpan.Length)
+            {
+                GetNextSpan();
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Move the reader back the specified number of positions.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Rewind(long count)
+        {
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
+            }
+
+            Consumed -= count;
+
+            if (CurrentSpanIndex >= count)
+            {
+                CurrentSpanIndex -= (int)count;
+                _moreData = true;
+            }
+            else
+            {
+                // Current segment doesn't have enough space, scan backward through segments
+                RetreatToPreviousSpan(Consumed);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void RetreatToPreviousSpan(long consumed)
+        {
+            ResetReader();
+            Advance(consumed);
+        }
+
+        private void ResetReader()
+        {
+            CurrentSpanIndex = 0;
+            Consumed = 0;
+            _currentPosition = Sequence.Start;
+            _nextPosition = _currentPosition;
+
+            if (Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
+            {
+                _moreData = true;
+
+                if (memory.Length == 0)
+                {
+                    CurrentSpan = default;
+                    // No space in the first span, move to one with space
```

Comment:
By "space", this means "data available to read"?  I normally think of the term "space" being used when in the context of a writer looking for a place to store data.

Reply:
No data. I'll update the comments here.